src/diffusers/loaders.py

```diff
@@ -241,3 +243,172 @@ def save_attn_procs(
         save_function(state_dict, os.path.join(save_directory, weights_name))
 
         logger.info(f"Model weights saved in {os.path.join(save_directory, weights_name)}")
+
+
+class TextualInversionLoaderMixin:
+    r"""
+    Mixin class for adding textual inversion tokens and embeddings to the tokenizer and text encoder with method:
+    - [`~TextualInversionLoaderMixin.load_textual_inversion_embeddings`]
+    - [`~TextualInversionLoaderMixin.add_textual_inversion_embedding`]
+    """
+
+    def load_textual_inversion_embeddings(
+        self, embedding_path_dict_or_list: Union[Dict[str, str], List[Dict[str, str]]], allow_replacement: bool = False
+    ):
+        r"""
+        Loads textual inversion embeddings and adds them to the tokenizer's vocabulary and the text encoder's
+        embeddings.
+
+        Arguments:
+            embeddings_path_dict_or_list (`Dict[str, str]` or `List[str]`):
+                Dictionary of token to embedding path or List of embedding paths to embedding dictionaries. The
+                dictionary must have the following keys:
+                    - `token`: name of the token to be added to the tokenizers' vocabulary
+                    - `embedding`: path to the embedding of the token to be added to the text encoder's embedding
+                      matrix
+                The list must contain paths to embedding dictionaries where the keys are the tokens and the values are
+                the embeddings (same as above dictionary definition).
+
+        Returns:
+            None
+        """
+        # Validate that inheriting class instance contains required attributes
+        self._validate_method_call(self.load_textual_inversion_embeddings)
+
+        if isinstance(embedding_path_dict_or_list, dict):
+            for token, embedding_path in embedding_path_dict_or_list.items():
+                # check if token in tokenizer vocab
+                if token in self.tokenizer.get_vocab():
+                    if allow_replacement:
+                        logger.info(
+                            f"Token {token} already in tokenizer vocabulary. Overwriting existing token and embedding"
+                            " with the new one."
+                        )
+                    else:
+                        raise ValueError(
+                            f"Token {token} already in tokenizer vocabulary. Please choose a different token name."
+                        )
+
+                embedding_dict = torch.load(embedding_path, map_location=self.text_encoder.device)
```

Comment:
We have a utility function `diffusers.models.modeling_utils` that can support loading safetensors as well :)

Reply:
This is great! 

Slightly off-topic: I have a working implementation of this feature in a forked branch with some additional modifications for loading the files directly from s3. Does diffusers already have integrations for loading from remote file store like s3? Temporarily saving to local disk first is expensive in production, especially where you might need to load different inversions on every call.

I know HF has been deepening partnerships with various cloud providers and am curious where you're trying to draw the line in terms of feature coupling.

Reply:
For storage, we heavily rely on the HF Hub. Also, we aim to keep diffusers to be light in terms of dependencies. So, s3 is probably a bit out of scope for now, but I will let @patrickvonplaten also share his thoughts.